### ISSUE TITLE
Demystifying "steps"

### ISSUE THREAD

**Issue by danyaljj** [google-research/text-to-text-transfer-transformer#174]
Hello, 
I wanted to get some clarity as to what "step" means.
In particular, I am interested in learning how "step" is related to the number of observed training data.
Is it fair to say that: `# of steps x batch-size` equals the number of observed training instances? 
If so, after  `training-data-size / (# of steps x batch-size)` many steps, we should have seen all instances in the training data.

**Comment by craffel** [comment]
Hi Daniel,

> I wanted to get some clarity as to what "step" means.

A step is a gradient update, nothing more, nothing less.

> Is it fair to say that: # of steps x batch-size equals the number of observed training instances?

Not exactly, because of [packing](https://github.com/tensorflow/mesh/blob/master/mesh_tensorflow/transformer/dataset.py#L16). It's easier to think about how many packed *tokens* are in each batch (including padding). We usually specify e.g. "65536 tokens per batch with a sequence length of 512". This means that there are 128 "entries" in the batch but each "entry" can have multiple examples because of packing. If you want to know how many steps corresponds to an epoch, you need to know how efficient packing is (i.e. how much of the batch is actual data vs. just padding symbols; maybe 80% depending on various factors including the packed sequence length and the distribution of sequence lengths of examples in your dataset). Usually I just do a guesstimate where I say "dataset has N training tokens, I have M tokens in each batch, and my batches are 80% train tokens, so an epoch is about N/(M*0.8)".

You can avoid this fudginess if you don't pack data and just use one example per entry in the batch (with padding), but packing can provide a huge boost in efficiency so really should be used.

**Comment by danyaljj** [comment]
Makes sense, thank you!
